assembly/runtime/promise.ts

```diff
@@ -0,0 +1,145 @@
+import { util } from "./util";
+import { env } from "./env";
+import { u128 } from ".";
+
+function encodeArgs<T>(t: T) : Uint8Array {
+  if (t instanceof Uint8Array) {
+    return t;
+  }
+  return encode<T>(t);
+}
+
+// TODO: not sure if these types make any sense at all
+export type Balance = u128;
+export type Gas = u64;
+export type PublicKey = Uint8Array;
+export type AccountId = Uint8Array;
+
+export type CreateAccount = "CreateAccount"; // TODO: this seems wierd
+export type DeployContract = { code: Uint8Array };
+export type FunctionCall = { methodName: Uint8Array, args: Uint8Array, amount: Balance, gas: Gas };
+export type Transfer = { amount: Balance };
+export type Stake = { amount: Balance, publicKey: PublicKey };
+export type AddFullAccessKey = { publicKey: PublicKey };
+export type AddAccessKey = { publicKey: PublicKey, allowance: Balance, receiverId: AccountId, methodNames: Uint8Array[] };
+export type DeleteKey = { publicKey: PublicKey };
+export type DeleteAccount = { beneficiaryId: AccountId };
+
+export type Action =
+  | CreateAccount
+  | DeployContract
+  | FunctionCall
+  | Transfer
+  | Stake
+  | AddFullAccessKey
+  | AddAccessKey
+  | DeleteKey
+  | DeleteAccount;
+
+/**
+ * Chain actions
+ *
+ * let p1 = new ActionPromise("bob_near").create_account();
+ * let p2 = new ActionPromise("alice_near").create_account().transfer(100);
+ */
+export class ContractPromiseBatch {
+  actions: Action[];
+  after: ContractPromiseBatch;
+  id: u64;
+
+  static create(
+    accountId: string
+  ) : ContractPromiseBatch {
+    const accountIdArr = util.stringToBytes(accountId);
+    const id: u64 = env.promise_batch_create(accountIdArr.byteLength, accountIdArr.dataStart);
+
+    //@ts-ignore: Typescript expects the object to have a function to match the type
+    // Wheras as only cares about the fields.
+    return {
+      id,
+    };
+  }
+
+  add_action(action: Action) : void {
+    this.actions.push(action);
+  }
+
+  create_account() : ContractPromiseBatch {
+    env.promise_batch_action_create_account(this.id);
+    this.add_action("CreateAccount");
+    return this;
+  }
+
+  deploy_contract(code: Uint8Array) : ContractPromiseBatch {
+    env.promise_batch_action_deploy_contract(this.id, code.length, code.dataStart);
+    this.add_action({ code });
+    return this;
+  }
+
+  // TODO: why is amount: Balance a u128 even though the batch_action interface takes a u64?  seems like we're losing half the data here. same happens above w ContractPromise btw
```

Comment:
Isn't it a u64 pointer and u64 length?

Reply:
still to discuss, not clear on the types here whether `u128` or `u64`

Reply:
The u128 is converted to an array and the pointer to that array is `u64` and that is passed.  You don't need to provide the length since a u128 is always 16 bytes.